calc/src/calc_partial_fee.rs

```diff
@@ -0,0 +1,259 @@
+    /// it and seeing whether the node changes the weight/decides not to take a fee at all.
+    ///
+    /// To get the correct values for some extrinsic from both endpoints, provide the extrinsic bytes, and the
+    /// block number **one before the block it made it into** (eg if the extrinsic was in block 100, you'd use
+    /// block 99 as an argument). This is very important.
+    ///
+    /// Once you've called these endpoints, access the `ExtrinsicSuccess` event to find the `actual_weight`, but
+    /// also a `paysFee` value which signals whether the extrinsic actually incurred a fee at all or not (a node
+    /// has the opportunity to refund the fee entirely if it likes by setting this).
+    ///
+    /// With all of those values to hand, the equation above calculates the correct Fee. Why? Well, the basic
+    /// way to calculate a pre-dispatch fee is:
+    ///
+    /// ```
+    /// partial_fee = base_fee + len_fee + adjusted_weight_fee
+    /// ```
+    ///
+    /// We can do this from just the RPC methods. But then once it's in a block, we need to swap out the weight used
+    /// to calculate that `adjusted_weight_fee` with the actual weight that was used from the `ExtrinsicSuccess` event.
+    /// In the end, the maths is simple and gathering the details needed is the main difficulty. We do this all in
+    /// Rust simply to limit any precision loss.
+    pub fn calc(
+        base_fee: &str,
+        len_fee: &str,
+        adjusted_weight_fee: &str,
+        estimated_weight: &str,
+        actual_weight: &str
+    ) -> CalcPartialFee {
+        let base_fee = new_u128(base_fee);
+        let len_fee = new_u128(len_fee);
+        let adjusted_weight_fee = new_u128(adjusted_weight_fee);
+        let estimated_weight = new_u128(estimated_weight);
+        let actual_weight = new_u128(actual_weight);
+
+        let partial_fee = Self::calc_raw(
+            base_fee,
+            len_fee,
+            adjusted_weight_fee,
+            estimated_weight,
+            actual_weight
+        );
+
+        Self {
+            partial_fee
+        }
+    }
+
+    /// Return the calculated partial fee as a string, to avoid any precision
+    /// loss as it crosses the boundary back to JS land.
+    pub fn partial_fee_as_string(&self) -> String {
+        self.partial_fee.to_string()
+    }
+
+    fn calc_raw(
+        base_fee: u128,
+        len_fee: u128,
+        adjusted_weight_fee: u128,
+        estimated_weight: u128,
+        actual_weight: u128
+    ) -> u128 {
+        // calculate new adjusted_weight_fee, trying to maintain precision:
+        let adjusted_weight_fee = Perbill::from_rational(estimated_weight, actual_weight) * adjusted_weight_fee;
+
+        // add the fee components together to get the partial/inclusion fee:
+        base_fee.saturating_add(len_fee).saturating_add(adjusted_weight_fee)
+    }
+}
+
+fn new_u128(inner: &str) -> u128 {
+    inner.parse().expect("string is not a valid u128")
```

Comment:
is this an exception o what happens when this fails when invoked from JS land?

Reply:
Def checking this @TarikGul. In fact, I have a vague feeling that `Result<Foo,Bar>` is properly handled nowadays with WASM so it might be worth moving to do it that way instead, so that it's more "rusty".

Reply:
(This was partly my fault too; I threw a load of unwraps in and didn't think too much about it!)

Reply:
Resolved. We made everything top level functions, and used a Result<String, JsError> type to error handle.